includes/sanitizers/class-amp-core-theme-sanitizer.php

```diff
@@ -1932,6 +1998,257 @@ public function add_twentytwenty_toggles() {
+
+		// Create an <amp-lightbox> element that will contain the modal.
+		$amp_lightbox = $this->dom->createElement( 'amp-lightbox' );
+		$amp_lightbox->setAttribute( 'layout', 'nodisplay' );
+		$amp_lightbox->setAttribute( 'animate-in', 'fade-in' );
+		$amp_lightbox->setAttribute( 'scrollable', true );
+		$amp_lightbox_id = $this->dom->getElementId( $amp_lightbox );
+
+		$state_string = str_replace( '-', '_', $amp_lightbox_id );
+
+		AMP_DOM_Utils::add_amp_action( $menu_toggle, 'tap', "{$amp_lightbox_id}.open" );
+		AMP_DOM_Utils::add_amp_action( $menu_toggle, 'tap', "{$body_id}.toggleClass(class=primary-navigation-open,force=true)" );
+
+		AMP_DOM_Utils::add_amp_action( $amp_lightbox, 'lightboxOpen', "AMP.setState({{$state_string}:true})" );
+		AMP_DOM_Utils::add_amp_action( $amp_lightbox, 'lightboxClose', "AMP.setState({{$state_string}:false})" );
+		AMP_DOM_Utils::add_amp_action( $amp_lightbox, 'lightboxClose', "{$body_id}.toggleClass(class=primary-navigation-open,force=false)" );
+
+		$menu_toggle->setAttribute( 'data-amp-bind-aria-expanded', "{$state_string} ? 'true' : 'false'" );
+
+		$amp_lightbox->appendChild( $primary_menu_copy );
+		$primary_menu->parentNode->insertBefore( $amp_lightbox, $primary_menu );
+	}
+
+	/**
+	 * Make the sub-menu functionality for the Twenty Twenty-One theme AMP compatible.
+	 *
+	 * Note: Hover functionality is accomplished through CSS.
+	 *
+	 * @see amend_twentytwentyone_styles()
+	 */
+	public function add_twentytwentyone_sub_menu_fix() {
+		// The XPath query has to be specific enough to not interfere with the mobile menu.
+		$menu_toggles = $this->dom->xpath->query( '//nav/div/ul//button[ @class and contains( concat( " ", normalize-space( @class ), " " ), " sub-menu-toggle " ) ]' );
+
+		if ( 0 === $menu_toggles->length ) {
+			return;
+		}
+
+		$menu_toggle_ids = substr_replace( range( 1, $menu_toggles->length ), 'toggle_', 0, 0 );
+
+		// Sub-menus to be closed when the user clicks on the body.
+		$toggles_to_disable_for_body = [];
+
+		foreach ( $menu_toggle_ids as $key => $menu_toggle_id ) {
+			/** @var DOMElement $menu_toggle */
+			$menu_toggle = $menu_toggles->item( $key );
+
+			$menu_toggle->setAttribute( 'data-amp-bind-aria-expanded', "{$menu_toggle_id} ? 'true' : 'false'" );
+
+			// Sub-menus to be closed when this one is to be opened.
+			$toggles_to_disable            = '';
+			$toggles_to_disable_for_body[] = "{$menu_toggle_id}:false";
+
+			foreach ( $menu_toggle_ids as $other_menu_toggle_id ) {
+				if ( $menu_toggle_id === $other_menu_toggle_id ) {
+					continue;
+				}
+
+				$toggles_to_disable .= ",{$other_menu_toggle_id}:false";
+			}
+
+			AMP_DOM_Utils::add_amp_action( $menu_toggle, 'tap', "AMP.setState({{$menu_toggle_id}:!{$menu_toggle_id}{$toggles_to_disable}})" );
+		}
+
+		$state_vars = implode( ',', $toggles_to_disable_for_body );
+		AMP_DOM_Utils::add_amp_action( $this->dom->body, 'tap', "AMP.setState({{$state_vars}})" );
```

Comment:
One tiny drawback I'm seeing with this approach is that when the `body` has a tap handler, `AMP_Accessibility_Sanitizer` adds `role="button"` and `tab-index="0"`, effectively turning the entire body into a button. I think this will confuse screen readers a bit (and cause the Skip to Content link to come second instead of first). Looks like `AMP_Accessibility_Sanitizer` would skip the body element if we set a tab index and a role. We could set tabindex to "-1" and explicitly set a role. Not sure [which role is best](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles), though -- maybe `document`?

Reply:
@westonruter do you think it would be better to introduce a filter in `AMP_Accessibility_Sanitizer` to exclude the `body`?

Reply:
The `role` and `tabindex` attributes are required by AMP for any element that has the `on` attribute:

![image](https://user-images.githubusercontent.com/134745/101558763-7e20cb00-3974-11eb-9526-99144f68cd75.png)

Reply:
This seems to work fine:

```suggestion
		AMP_DOM_Utils::add_amp_action( $this->dom->body, 'tap', "AMP.setState({{$state_vars}})" );
		$this->dom->body->setAttribute( 'role', 'document' );
		$this->dom->body->setAttribute( 'tabindex', '-1' );
```

Reply:
Could we move the `amp-bind` to `#page` instead and use `role="main"`? That seems like a more accurate role.

Reply:
Ah nevermind, it wouldn't be.

Reply:
Alternatively, we could come up with a custom role value. This is not validated.

Reply:
@westonruter applied your suggestion in https://github.com/ampproject/amp-wp/pull/5648/commits/850e530c334b8070a0098529d5e68d81e7897ebf.